### PR TITLE
respect ExecProbeTimeout=false for dockershim

### LINKED ISSUE #100198: exec probes that exceed timeout always succeed when using ExecProbeTimeout=false
#### What happened:

We've observed that when using a `ExecProbeTimeout=false` kubelet feature gate configuration with dockershim (e.g., docker/moby CRI) that exec probes have no effect if their command runtime takes longer than the configured probe `timeoutSeconds`.

This is different than the buggy behavior of dockershim probe timeout prior to the release of 1.20. That previous behavior, which `ExecProbeTimeout=false` is meant to reproduce in 1.20 and beyond, simply ignored timeouts, allowing probes "unlimited" time to complete, and respecting the eventual probe command exit code.

What we are observing now: Instead of timeouts being entirely ignored when using `ExecProbeTimeout=false`, a probe exceeding the timeout has the effect of short-circuiting the context between the probe and the container runtime entirely, and thus probes seem to "always succeed". What's really happening is that the outcome of the probe, after its time duration exceeds the configured timeout (1 second by default), is ignored. The container continues running.

#### What you expected to happen:

I would expect for a probe command that took longer than the configured timeout to still be operative, and for a non-zero probe command outcome to result in the container being killed, and the restart property incremented by 1.

#### How to reproduce it (as minimally and precisely as possible):

1. Create a Kubernetes cluster >= 1.20.0 with a kubelet feature gate configuration that includes `ExecProbeTimeout=false`. (we tested in Azure using a docker CRI)
2. Launch a container that runs forever, for example a simple `busybox` Pod spec with a command like `while true; do sleep 5; done`, and include an exec livenessProbe that does something that will always return non-zero, like `sleep 10 & exit 1`; the default 1 second timeout will ensure that a probe command such as that one always exceeds the timeout; and the `ExecProbeTimeout=false` kubelet feature gate config will ensure that that timeout is ignored
3. Observe that that busybox container running the above spec will not restart; thus, the probe failure is not being respected

#### Anything else we need to know?:

@chewong created a simple test case that proves this here:

https://github.com/chewong/kubernetes/commit/e3632dd0183bbd9d5eb0789aaf4a16bac098e9e5

#### Environment:
- Kubernetes version (use `kubectl version`): 1.20.0 and up
- Cloud provider or hardware configuration: Azure
- OS (e.g: `cat /etc/os-release`):
```
NAME="Ubuntu"
VERSION="18.04.5 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.5 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```
- Kernel (e.g. `uname -a`): Linux k8s-master-46627706-0 5.4.0-1040-azure #42~18.04.1-Ubuntu SMP Mon Feb 8 19:05:32 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
- Install tools: Cluster built w/ [aks-engine](https://github.com/Azure/aks-engine)
- Network plugin and version (if this is a network-related bug): n/a
- Others:

### PR DESCRIPTION
**Pull request by jackfrancis**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

This PR ensures that the kubelet ExecProbeTimeout feature gate is respected when evaluating whether to timeout an exec probe's context when its configured `timeoutSeconds` period has expired.

The current implementation has the effect of ignoring the probe outcome after the `timeoutSeconds` period has expired.

#### Which issue(s) this PR fixes:

Fixes #100198

#### Special notes for your reviewer:

I've tested that a build with this PR restored the expected behavior, where timeouts are simply ignored.

The original PR that fixed the buggy timeout behavior *does not* reproduce this new "probes always succeed pathology" according to my tests:

https://github.com/kubernetes/kubernetes/pull/94115

Further testing confirmed that this commit is the first that reproduces this new bug:

https://github.com/kubernetes/kubernetes/pull/96495

It appears that this surface area change ommitted the ExecProbeTimeout feature gate:

https://github.com/kubernetes/kubernetes/pull/96495/files#diff-de4678da3e9d01e8ee0aadd641deb2ffec3a53c8603476509e59fa8dc5449deaL117

#### Does this PR introduce a user-facing change?

```release-note
respect ExecProbeTimeout=false for dockershim
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by jackfrancis** [comment]
cc @andrewsykim @chewong @lachie83 

It wasn't obvious to me how me might inject a feature gate config into the mocks used by `TestExecInContainer` in `pkg/kubelet/dockershim/exec_test.go`. If we can figure that out it should be easy to add a new test case to this PR that proves the fix.

Additionally, do you think it would make sense to include something like this in this PR?:

https://github.com/chewong/kubernetes/commit/b59d9039ee4bffbca7d7360325911f125c20334f

File: test/e2e/common/node/container_probe.go [hunk 1]
```diff
@@ -229,6 +229,25 @@ var _ = SIGDescribe("Probing container", func() {
 		RunLivenessTest(f, pod, 1, defaultObservationTimeout)
 	})
 
+	/*
+		Release: v1.20
+		Testname: Pod liveness probe, container exec timeout, restart
+		Description: A Pod is created with liveness probe with a Exec action on the Pod. If the liveness probe call does not return within the timeout specified, liveness probe MUST restart the Pod. When ExecProbeTimeout feature gate is disabled, the timeout is ignored BUT a failing liveness probe MUST restart the Pod.
+	*/
+	ginkgo.It("should be restarted with a failing exec liveness probe that took longer than the timeout [NodeConformance]", func() {
```
Comment (ehashman) [inline test/e2e/common/node/container_probe.go:237]: We can't add an e2e test directly to conformance suites. They need to run in CI for at least two weeks first.

**Review by ehashman — COMMENTED** [review]
/priority important-soon
/skip

File: test/e2e/common/node/container_probe.go (hunk 1 above)
Comment (jackfrancis) [reply test/e2e/common/node/container_probe.go:237]: Got it. @chewong You want to split this off into a separate PR?

Comment (chewong) [reply test/e2e/common/node/container_probe.go:237]: we can just remove the [NodeConformance] tag from the spec

Comment (jackfrancis) [reply test/e2e/common/node/container_probe.go]: O.K., removed the [NodeConformance] tag

Comment (jackfrancis) [reply test/e2e/common/node/container_probe.go]: @ehashman lemme know removing `[NodeConformance]` was the right resolution here, thank you!

File: pkg/kubelet/dockershim/exec.go [hunk 2]
```diff
@@ -106,7 +108,7 @@ func (*NativeExecHandler) ExecInContainer(ctx context.Context, client libdocker.
 		ExecStarted:  execStarted,
 	}
 
-	if timeout > 0 {
+	if timeout > 0 && utilfeature.DefaultFeatureGate.Enabled(features.ExecProbeTimeout) {
```
Comment (andrewsykim) [inline pkg/kubelet/dockershim/exec.go:111]: The thinking here was that the timeout is always respected, but the kubelet just ignores it from the prober:

https://github.com/kubernetes/kubernetes/blob/4bb30c3b0e9167eb85b1e3b0c9a3bab3b351a3e2/pkg/probe/exec/exec.go#L72-L79

Is that behavior not happening?

Comment (jackfrancis) [reply pkg/kubelet/dockershim/exec.go:111]: Also, again: we've only observed this `ExecProbeTimeout=false` pathology in dockershim-enabled clusters (hence the change in this surface are). So this is a docker-only thing as far as we've discovered.

Comment (andrewsykim) [reply pkg/kubelet/dockershim/exec.go:111]: > What's happening in the ExecProbeTimeout=false configuration is that (1) the timeout being reached does not initiate a container reset and (2) if the actual exec command fails after the configured timeout configuration then similarly the container reset does not happen.

Sorry I think I'm still missing context -- I _think_ this the old behavior we're trying to preserve? Prior to implementing exec probe timeouts, this was the behavior from dockershim:
* ignore probe timeouts entirely, where timedout liveness probe do not restart containers
* if the exec probe fails after the timeout (or after 10 seconds), it is not evaluated from the prober

Comment (jackfrancis) [reply pkg/kubelet/dockershim/exec.go:111]: Agree 100% the intent is to preserve the pre-1.20.0 dockershim behavior.

What we see doesn't match the 2nd bullet point above. If the exec probe fails after the timeout, it *is* evaluated from the prober.

I have a PR in the aks-engine project that shows this happening:

https://github.com/Azure/aks-engine/pull/4325

Basically, the new tests pass in 1.18 and 1.19, but fail in 1.20. The one that's important is this one:

https://github.com/Azure/aks-engine/pull/4325/files#diff-0bd2bb48395375d835916863ebd4b2bf05747e18d16d8e551588bec8a447b470

The above spec checks out in 1.18 and 1.19: pod restarts are observed on 5+ second failures despite the fact that the 1 second timeout is "ignored".

Comment (andrewsykim) [reply pkg/kubelet/dockershim/exec.go:111]: > What we see doesn't match the 2nd bullet point above. If the exec probe fails after the timeout, it is evaluated from the prober.

Ah, this is the part I was missing :) So a probe failing after the timeout does trigger a container restart when ExecProbeTimeout=false?

Comment (jackfrancis) [reply pkg/kubelet/dockershim/exec.go:111]: We've tested 1.18 and 1.19 (there is no ExecProbeTimeout feature flag there). The behavior in 1.18 and 1.19 is that a failing probe that takes longer than a timeout triggers a container restart. That's what we want 1.20+ and `ExecProbeTimeout=false` to replicate.

But what we see is that 1.20+ and `ExecProbeTimeout=false` do not restart containers when the probe fails after the timeout period.

tl;dr for dockershim: (pre-1.20) compared to (1.20 + `ExecProbeTimeout=false`) are not equivalent.

Comment (andrewsykim) [reply pkg/kubelet/dockershim/exec.go:111]: Gotcha, thanks for clarifying.

> The behavior in 1.18 and 1.19 is that a failing probe that takes longer than a timeout triggers a container restart.

One clarifying question on this, do you know if this is true for timeouts greater than 10 seconds?

Comment (jackfrancis) [reply pkg/kubelet/dockershim/exec.go:111]: I'll test and report back!

Comment (jackfrancis) [reply pkg/kubelet/dockershim/exec.go:111]: 10+ second probe failure restarts the container as expected in 1.19.8:

```
STEP: Validating that the exec livenessProbe caused at least one pod restart due to command failure, even if timeout is not respected
2021/03/29 13:15:43 $ k describe pod exec-liveness-timeout-always-fail -n default
2021/03/29 13:15:44 #### $ k describe pod exec-liveness-timeout-always-fail -n default completed in 256.907866ms
2021/03/29 13:15:44 
Name:         exec-liveness-timeout-always-fail
Namespace:    default
Priority:     0
Node:         k8s-agentpool1-28118473-vmss000001/10.240.0.65
Start Time:   Mon, 29 Mar 2021 13:14:56 -0700
Labels:       test=liveness
Annotations:  kubernetes.io/psp: privileged
Status:       Running
IP:           10.240.0.68
IPs:
  IP:  10.240.0.68
Containers:
  exec-liveness-timeout-always-fail:
    Container ID:  docker://6a9a11af16d487e3722cfa33e2565550f425f70d5c726da6d69b3e9bea337b5b
    Image:         busybox:1.31.1
    Image ID:      docker-pullable://busybox@sha256:95cf004f559831017cdf4628aaf1bb30133677be8702a8c5f2994629f637a209
    Port:          <none>
    Host Port:     <none>
    Args:
      /bin/sh
      -c
      while true; do rm -f /tmp/healthy; sleep 5; done
    State:          Running
      Started:      Mon, 29 Mar 2021 13:15:43 -0700
    Last State:     Terminated
      Reason:       Error
      Exit Code:    137
      Started:      Mon, 29 Mar 2021 13:15:28 -0700
      Finished:     Mon, 29 Mar 2021 13:15:43 -0700
    Ready:          True
    Restart Count:  3
    Liveness:       exec [sh -c sleep 11 && test -f /tmp/healthy] delay=1s timeout=1s period=1s #success=1 #failure=1
    Environment:    <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-fzsh5 (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             True 
  ContainersReady   True 
  PodScheduled      True 
Volumes:
  default-token-fzsh5:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-fzsh5
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  kubernetes.io/os=linux
Tolerations:     node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                 node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason     Age               From               Message
  ----     ------     ----              ----               -------
  Normal   Scheduled  47s               default-scheduler  Successfully assigned default/exec-liveness-timeout-always-fail to k8s-agentpool1-28118473-vmss000001
  Warning  Unhealthy  3s (x3 over 33s)  kubelet            Liveness probe failed:
  Normal   Killing    3s (x3 over 33s)  kubelet            Container exec-liveness-timeout-always-fail failed liveness probe, will be restarted
  Normal   Pulled     1s (x4 over 46s)  kubelet            Container image "busybox:1.31.1" already present on machine
  Normal   Created    1s (x4 over 46s)  kubelet            Created container exec-liveness-timeout-always-fail
  Normal   Started    0s (x4 over 46s)  kubelet            Started container exec-liveness-timeout-always-fail
```

Comment (liggitt) [reply pkg/kubelet/dockershim/exec.go:111]: The current change looks good to me for preserving pre-1.20 behavior. Rather than checking for TimeoutError now to warn about exceeding the timeout, should we now check the actual time it took to execute the probe and warn if that exceeded the configured timeout?

As a follow-up, in 1.22 (and maybe picked back to 1.21 / 1.20), I'd actually like to see a metric published for probes whose execution time exceeds their timeout (or maybe exceeds 90% of the timeout or something) so admins can determine whether it is safe to set ExecProbeTimeout=true.

File: pkg/kubelet/dockershim/exec_test.go [hunk 3]
```diff
@@ -110,23 +143,27 @@ func TestExecInContainer(t *testing.T) {
 	var resize <-chan remotecommand.TerminalSize
 
 	for _, tc := range testcases {
-		t.Logf("TestCase: %q", tc.description)
+		tc := tc
```
Comment (andrewsykim) [inline pkg/kubelet/dockershim/exec_test.go:146]: Why is this line needed?

Comment (jackfrancis) [reply pkg/kubelet/dockershim/exec_test.go:146]: It has to do with the closure provided by the enumeration (`tc` is closed over inside the `testcases` slice) and how the invocation of test.Parallel() can produce variable reference collisions.

This public gist is a nice explanation: https://gist.github.com/posener/92a55c4cd441fc5e5e85f27bca008721#what-happened

However! We're not actually invoking `.Parallel()` :/

Should we run these tests concurrently? If so, I'll add the `.Parallel()` foo, otherwise, yeah, I'll remove that `tc := tc` (it's probably just a habit from writing parallel golang UT).

Comment (jackfrancis) [reply pkg/kubelet/dockershim/exec_test.go:146]: I went ahead and enabled parallelization after doing a cursory audit of the UT landscape. As you'd expect in k/k, there are a million different patterns in the codebase, but I saw enough of these that I felt that the way we're introducing parallelization here is not without precedent. E.g.:

- https://github.com/kubernetes/kubernetes/blob/master/pkg/security/podsecuritypolicy/util/util_test.go#L462
- https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/runtime/converter_test.go#L586
- https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/watch/retrywatcher_test.go#L514

(etc)

File: test/e2e/common/node/container_probe.go [hunk 4]
```diff
@@ -246,6 +246,25 @@ var _ = SIGDescribe("Probing container", func() {
 		runReadinessFailTest(f, pod, time.Minute)
 	})
 
+	/*
+		Release: v1.21
+		Testname: Pod liveness probe, container exec timeout, restart
+		Description: A Pod is created with liveness probe with a Exec action on the Pod. If the liveness probe call does not return within the timeout specified, liveness probe MUST restart the Pod. When ExecProbeTimeout feature gate is disabled, the timeout is ignored BUT a failing liveness probe MUST restart the Pod.
```
Comment (SergeyKanzhelev) [inline test/e2e/common/node/container_probe.go:252]: this is not the behavior on containerd if the timeout is higher than the default grpc timeout, correct? Should this be mentioned here?

Comment (SergeyKanzhelev) [inline test/e2e/common/node/container_probe.go:252]: will this change how Containerd behaves? Before, in 1.19, the timeout wasn't respected and the probe failing after the timeout, will not restart the pod. Is it intentional to fix this behavior?

Comment (jackfrancis) [reply test/e2e/common/node/container_probe.go:252]: I think the easiest way to think of this is that `ExecProbeTimeout=false` is a post-1.20 solution that is only appropriate for dockershim configuration.

This PR does not intend to fix the containerd scenario. For containerd scenarios (all non-dockershim scenarios, probably), we should not recommend fiddling with ExecProbeTimeout.

Should we make that more clear in the comments?

File: pkg/kubelet/dockershim/exec_test.go [hunk 5]
```diff
@@ -110,23 +143,28 @@ func TestExecInContainer(t *testing.T) {
...
+		t.Run(tc.description, func(t *testing.T) {
+			t.Parallel()
+			defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.ExecProbeTimeout, tc.execProbeTimeout)()
+			t.Logf("TestCase: %q", tc.description)
```
Comment (andrewsykim) [inline pkg/kubelet/dockershim/exec_test.go:150]: nit: This log is not needed since t.Run would print the description already

File: test/e2e/common/node/container_probe.go (hunk 4 above)
Comment (andrewsykim) [reply test/e2e/common/node/container_probe.go:252]: > Should we make that more clear in the comments?

Yes, I think so

File: test/e2e/common/node/container_probe.go [hunk 6]
```diff
@@ -246,6 +246,25 @@ var _ = SIGDescribe("Probing container", func() {
...
+	*/
+	ginkgo.It("should be restarted with a failing exec liveness probe that took longer than the timeout", func() {
+		// The ExecProbeTimeout feature gate exists to allow backwards-compatibility with pre-1.20 cluster behaviors, where livenessProbe timeouts were ignored
```
Comment (andrewsykim) [inline test/e2e/common/node/container_probe.go:255]: You may have mentioned this already, so sorry for asking again. What's the reason for not conditionally running this test when ExecProbeTimeout=false? Is it because the test would pass regardless because of the timeout?

File: test/e2e/common/node/container_probe.go (hunk 4 above)
Comment (SergeyKanzhelev) [reply test/e2e/common/node/container_probe.go:252]: yes, comment will suffice. So in 1.20 the behavior for containerd has changed - now it respects the failed probe after timeout. This is a breaking change, but I'd say it is not that critical.

File: test/e2e/common/node/container_probe.go [hunk 7]
```diff
@@ -246,6 +246,26 @@ var _ = SIGDescribe("Probing container", func() {
 		runReadinessFailTest(f, pod, time.Minute)
 	})
 
+	/*
+		Release: v1.21
+		Testname: Pod liveness probe, container exec timeout, restart
+		Description: A Pod is created with liveness probe with a Exec action on the Pod. If the liveness probe call does not return within the timeout specified, liveness probe MUST restart the Pod. When ExecProbeTimeout feature gate is disabled and cluster is using dockershim, the timeout is ignored BUT a failing liveness probe MUST restart the Pod.
+	*/
+	ginkgo.It("should be restarted with a failing exec liveness probe that took longer than the timeout", func() {
+		// The ExecProbeTimeout feature gate exists to allow backwards-compatibility with pre-1.20 cluster behaviors using dockershim, where livenessProbe timeouts were ignored
```
Comment (jackfrancis) [inline test/e2e/common/node/container_probe.go:255]: @andrewsykim @SergeyKanzhelev are these comment changes an improvement?

File: test/e2e/common/node/container_probe.go [hunk 8]
```diff
@@ -246,6 +246,26 @@ var _ = SIGDescribe("Probing container", func() {
 		runReadinessFailTest(f, pod, time.Minute)
 	})
 
+	/*
+		Release: v1.21
+		Testname: Pod liveness probe, container exec timeout, restart
+		Description: A Pod is created with liveness probe with a Exec action on the Pod. If the liveness probe call does not return within the timeout specified, liveness probe MUST restart the Pod. When ExecProbeTimeout feature gate is disabled and cluster is using dockershim, the timeout is ignored BUT a failing liveness probe MUST restart the Pod.
+	*/
+	ginkgo.It("should be restarted with a failing exec liveness probe that took longer than the timeout", func() {
```
Comment (jackfrancis) [inline test/e2e/common/node/container_probe.go:254]: @andrewsykim You asked why we aren't running this test in an ExecProbeTimeout=false scenario only. The main reason is that not all ExecProbeTimeout=false scenarios will pass this new test (e.g., containerd). I think the value of this new test is to guarantee that a failing probe fails *no matter what*. Since we have tests that validate timeouts fire as expected, this test should in practice be a reproduction of that pod restart condition (pod restarted due to timeout), but this additional test coverage will ensure that certain pre-1.20 behaviors (containerd, for example) will be caught right away in E2E tests if they resurface for any weird reason.

**Comment by SergeyKanzhelev** [comment]
/lgtm 

as is. Let's try to hit the Friday deadline for 1.20 cherry-pick?

**Comment by dims** [comment]
and 1.21 too @SergeyKanzhelev ? :)

**Comment by SergeyKanzhelev** [comment]
@andrewsykim any concerns?

**Comment by SergeyKanzhelev** [comment]
> and 1.21 too @SergeyKanzhelev ? :)

there is no 1.21 patch release, right? I meant I hope to hit the deadline for this 1.20 patch

**Review by hasheddan — APPROVED** [review]
/lgtm

**Review by ehashman — APPROVED** [review]
/lgtm

**Comment by ehashman** [comment]
flake #100556
